### PR TITLE
feature/Ds-72-Create-and-implement-login-end-point-and-flow

### PR DESCRIPTION
**Pull request by M-Jhay**
…Api-key, Email and password combination, user status and generated access token

### CODE REVIEW

**Comment by M-Jhay** [comment]
Added some environment variable for authorization middleware class that validates the user's source Ip and API key
X_API_KEY = Api key
SOURCE_IP = IP address

File: src/DwitTech.AccountService.Core/Services/UserService.cs [hunk 1]
```diff
@@ -1,8 +1,66 @@
...
+
+        private static bool IsUserActivated(User user)
+        {
+            if (user.Status == Data.Enum.UserStatus.Active)
+            {
+                return true;
+            }
+            return false;
+        }
+        public async Task<TokenModel> UserLogin (string email, string hashedPassword)
```
Comment (giftedGeorge) [inline src/DwitTech.AccountService.Core/Services/UserService.cs]: The name "UserLogin" doesn't follow the recommended naming convention. The method should be named like a verb to clearly explain its function.

Comment (M-Jhay) [reply src/DwitTech.AccountService.Core/Services/UserService.cs]: > The name "UserLogin" doesn't follow the recommended naming convention. The method should be named like a verb to clearly explain its function.

Resolved

**Review by giftedGeorge — DISMISSED** [review]

**Review by johnny2565 — DISMISSED** [review]
The PR looks okay to me

File: src/DwitTech.AccountService.Data/Context/AccountDbContext.cs [hunk 2]
```diff
@@ -1,32 +1,42 @@
...
+        public AccountDbContext(DbContextOptions<AccountDbContext> options) : base(options)
+        {
+        }
+
+        public DbSet<User> Users { get; set; }
+        public DbSet<ValidationCode> ValidationCodes { get; set; }
+        public DbSet<Role> Roles { get; set; }
+        public DbSet<SessionToken> SessionTokens { get; set; }
+
+        public DbSet<UserLogin> UsersLogin { get; set; }
```
Comment (dabensymon) [inline src/DwitTech.AccountService.Data/Context/AccountDbContext.cs]:
```suggestion
        public DbSet<UserLogin> UserLogins { get; set; }
```

File: src/DwitTech.AccountService.WebApi/Controllers/UserController.cs [hunk 3]
```diff
@@ -1,32 +1,50 @@
...
+
+        }
+
+        [HttpPost]
+        [Route("login")]
+        public async Task<IActionResult> AuthenticateUserLogin([FromBody] string email, string hashedPassword)
+        {
+            try
+            {
+                var loginResult = await _userService.AuthenticateUserLogin(email, hashedPassword);
```
Comment (dabensymon) [inline src/DwitTech.AccountService.WebApi/Controllers/UserController.cs]: You aren't checking the login result if it is successful?

File: src/DwitTech.AccountService.Core/Services/UserService.cs [hunk 4]
```diff
@@ -1,8 +1,66 @@
...
+
+        private static bool IsUserActivated(User user)
+        {
+            if (user.Status == Data.Enum.UserStatus.Active)
+            {
+                return true;
+            }
+            return false;
+        }
+        public async Task<TokenModel> AuthenticateUserLogin (string email, string hashedPassword)
```
Comment (dabensymon) [inline src/DwitTech.AccountService.Core/Services/UserService.cs]: This should be in authentication service...

File: src/DwitTech.AccountService.Core/Services/UserService.cs [hunk 5]
```diff
@@ -1,8 +1,66 @@
...
+        {
+            var context = _httpContextAccessor.HttpContext;
+
+            //validate user's source Ip and Api Key
+            await _middleware.InvokeAsync(context);
```
Comment (dabensymon) [inline src/DwitTech.AccountService.Core/Services/UserService.cs]: No Way!!! Middleware is registered in the application pipeline. You can't invoke it here. Except if this validation only occurs for login. In that case, you dont need a midlleware.

File: src/DwitTech.AccountService.WebApi/Program.cs [hunk 6]
```diff
@@ -1,85 +1,96 @@
...
+            app.UseSwagger();
+            app.UseSwaggerUI();
+
+            app.UseForwardedHeaders();
+
+            app.UseHttpsRedirection();
+
+            app.UseRouting();
+
+            app.UseMiddleware<AuthorizationMiddleware>();
```
Comment (dabensymon) [inline src/DwitTech.AccountService.WebApi/Program.cs:78]: See? You have this middleware registered here. So no need to invoke it explicitly in the Service layer. Run the app and test your implementation.

**Review by dabensymon — CHANGES_REQUESTED** [review]
Good job. Just a couple of changes to be made

File: src/DwitTech.AccountService.Data/Context/AccountDbContext.cs (hunk 2 above)
Comment (M-Jhay) [reply src/DwitTech.AccountService.Data/Context/AccountDbContext.cs]: Th

> 

Resolved

Comment (M-Jhay) [reply src/DwitTech.AccountService.Data/Context/AccountDbContext.cs]: > 

Resolved

File: src/DwitTech.AccountService.WebApi/Controllers/UserController.cs (hunk 3 above)
Comment (M-Jhay) [reply src/DwitTech.AccountService.WebApi/Controllers/UserController.cs]: > You aren't checking the login result if it is successful?

Resolved

File: src/DwitTech.AccountService.Core/Services/UserService.cs (hunk 4 above)
Comment (M-Jhay) [reply src/DwitTech.AccountService.Core/Services/UserService.cs]: > This should be in authentication service...

Resolved

File: src/DwitTech.AccountService.Core/Services/UserService.cs (hunk 5 above)
Comment (M-Jhay) [reply src/DwitTech.AccountService.Core/Services/UserService.cs]: > No Way!!! Middleware is registered in the application pipeline. You can't invoke it here. Except if this validation only occurs for login. In that case, you dont need a midlleware.

Resolved

File: src/DwitTech.AccountService.WebApi/Program.cs (hunk 6 above)
Comment (M-Jhay) [reply src/DwitTech.AccountService.WebApi/Program.cs:78]: > See? You have this middleware registered here. So no need to invoke it explicitly in the Service layer. Run the app and test your implementation.

Noted boss

**Review by dabensymon — APPROVED** [review]

**Pull request merged**